plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/PyStringCodeCompletion.java

```diff
@@ -240,6 +245,88 @@ public TokensOrProposalsList getCodeCompletionProposals(CompletionRequest reques
         return ret;
     }
 
+    private TokensOrProposalsList getCompletionsForTypedDict(CompletionRequest request) throws CoreException,
+            BadLocationException, IOException, MisconfigurationException, PythonNatureWithoutProjectException {
+        if (request.isInKeytip) {
+            Optional<Tuple<String, String>> maybeActivationTokenAndQualifier = getActivationTokenAndQualifierStringsForDictKey(
+                    request);
+            if (maybeActivationTokenAndQualifier.isPresent()) {
+                Tuple<String, String> activationTokenAndQualifier = maybeActivationTokenAndQualifier.get();
+                String activationToken = activationTokenAndQualifier.o1;
+                String qualifier = activationTokenAndQualifier.o2;
+
+                CompletionRequest artificialRequest = new CompletionRequest(
+                        request.editorFile, request.nature, request.doc, activationToken,
+                        request.documentOffset, request.qlen, request.codeCompletion, qualifier,
+                        request.useSubstringMatchInCodeCompletion, request.isInKeytip);
+
+                return new PyCodeCompletion().getCodeCompletionProposals(artificialRequest);
+            }
+        }
+        return null;
+    }
+
+    final private static Optional<Tuple<String, String>> getActivationTokenAndQualifierStringsForDictKey(
+            CompletionRequest request)
+            throws BadLocationException {
+        Optional<Tuple<String, Integer>> maybeQualifierAndOffset = getQualifierAndOffsetForDictKey(request);
+        if (maybeQualifierAndOffset.isPresent()) {
+            Tuple<String, Integer> qualifierAndOffset = maybeQualifierAndOffset.get();
+            Optional<String> activationToken = getActivationTokenForDictKey(request, qualifierAndOffset.o2);
+            if (activationToken.isPresent()) {
+                return Optional.of(new Tuple<String, String>(activationToken.get(), qualifierAndOffset.o1));
+            }
+        }
+        return Optional.empty();
+    }
+
+    final private static Optional<Tuple<String, Integer>> getQualifierAndOffsetForDictKey(CompletionRequest request)
+            throws BadLocationException {
+        final IDocument doc = request.doc;
+        final FastPartitioner fastPartitioner = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc));
+
+        ITypedRegion partition = fastPartitioner.getPartition(request.documentOffset);
+        if (IPythonPartitions.PY_DEFAULT.equals(partition.getType())) { // string probably is open
+            for (int docOffset = request.documentOffset - 1; docOffset > 0; docOffset--) {
```

Comment:
It isn't ideal to iterate the whole document... just search backwards for a `["` or `['` (or until a character that doesn't match `Character.isJavaIdentifierPart`).

Reply:
I did an approach based on the search for char `'\'' || '"'`. Then the bracket check is done in the activationToken extraction.